Source/SwiftLintCore/Models/Baseline.swift

```diff
@@ -0,0 +1,186 @@
+        }
+
+        let violationsByRuleIdentifier = relativePathViolations.groupedByRuleIdentifier(
+            filteredBy: baselineViolations
+        )
+        let baselineViolationsByRuleIdentifier = baselineViolations.groupedByRuleIdentifier(
+            filteredBy: relativePathViolations
+        )
+
+        var filteredViolations: Set<BaselineViolation> = []
+
+        for (ruleIdentifier, ruleViolations) in violationsByRuleIdentifier {
+            guard
+                let baselineViolations = baselineViolationsByRuleIdentifier[ruleIdentifier],
+                    baselineViolations.isNotEmpty else {
+                filteredViolations.formUnion(ruleViolations)
+                continue
+            }
+
+            let groupedRuleViolations = Dictionary(grouping: ruleViolations, by: \.key)
+            let groupedBaselineViolations = Dictionary(grouping: baselineViolations, by: \.key)
+
+            for (key, ruleViolations) in groupedRuleViolations {
+                guard let baselineViolations = groupedBaselineViolations[key] else {
+                    filteredViolations.formUnion(ruleViolations)
+                    continue
+                }
+                if ruleViolations.count > baselineViolations.count {
+                    filteredViolations.formUnion(ruleViolations)
+                }
+            }
+        }
+
+        let violationsWithAbsolutePaths = Set(filteredViolations.violationsWithAbsolutePaths)
+        return violations.filter { violationsWithAbsolutePaths.contains($0) }
+    }
+
+    /// Returns the violations that are present in another `Baseline`, but not in this one.
+    ///
+    /// The violations are filtered using the same algorithm as the `filter` method above.
+    ///
+    /// - parameter otherBaseline: The other `Baseline`.
+    public func compare(_ otherBaseline: Baseline) -> [StyleViolation] {
+        otherBaseline.baselineViolations.flatMap {
+            filter($1.violationsWithAbsolutePaths)
+        }
+    }
+}
+
+private extension Sequence where Element == StyleViolation {
+    var baselineViolations: [BaselineViolation] {
+        var lines: [String: [String]] = [:]
+        var baselineViolations: [BaselineViolation] = []
+        for violation in self {
+            guard let absolutePath = violation.location.file,
+                  let lineNumber = violation.location.line != nil ? (violation.location.line ?? 0) - 1 : nil,
+                  lineNumber > 0 else {
+                baselineViolations.append(violation.baselineViolation())
+                continue
+            }
+            if let fileLines = lines[absolutePath] {
+                let text = (!fileLines.isEmpty && lineNumber < fileLines.count ) ? fileLines[lineNumber] : ""
+                baselineViolations.append(violation.baselineViolation(text: text))
+                continue
+            }
+            let text: String
+            if let fileLines = SwiftLintFile(path: absolutePath)?.lines.map(\.content),
```

Comment:
This is potentially time-consuming and memory-hungry as it's loading a whole project in the worst case ...

We may start with this, but I suspect that, in a next step, violations might get assigned the offending line upon creation or in a modification shortly after creation depending on the baseline functionality being active or not.

Reply:
So there definitely is an overhead - on my main project converting 78,785 violations (`--enable-all-rules`) in 2,197 files, c 300,000+ loc on my 16" 2021 MBP with an Apple M1 Max takes 1.5 - 2.5 seconds, so not great, but not terrible.

Just having a play around with memory now.

Reply:
So I definitely see a spike memory wise, but its not too massive compared to the rest of our consumption (with the same dataset as above).

<img width="1013" alt="Screen Shot 2024-04-15 at 22 20 28" src="https://github.com/realm/SwiftLint/assets/435558/aa765902-11e5-4510-87bc-f6c5489ee738">

Reply:
Okay. So let's go with the current implementation and optimize afterwards if necessary.